src/storage.rs

```diff
@@ -96,167 +151,230 @@ pub trait Storage {
+        self.raft_state.conf_states.clear();
+
+        let mut cs = ConfStateWithIndex::default();
+        cs.conf_state = meta.take_conf_state();
+        cs.index = meta.get_conf_state_index();
+        self.raft_state.conf_states.push(cs);
+
+        if meta.get_next_conf_state_index() > 0 {
+            let mut cs = ConfStateWithIndex::default();
+            cs.conf_state = meta.take_next_conf_state();
+            cs.index = meta.get_next_conf_state_index();
+            cs.in_membership_change = true;
+            self.raft_state.conf_states.push(cs);
         }
-        if let Some(pending_change) = pending_membership_change {
-            let meta = self.snapshot.mut_metadata();
-            meta.set_pending_membership_change(pending_change.get_configuration().clone());
-            meta.set_pending_membership_change_index(pending_change.get_start_index());
+    }
+
+    fn create_snapshot(&self) -> Snapshot {
+        let mut snapshot = Snapshot::new();
+
+        // Use the latest applied_idx to construct the snapshot.
+        let applied_idx = self.raft_state.hard_state.get_commit();
+        let term = self.raft_state.hard_state.get_term();
+        snapshot.mut_metadata().set_index(applied_idx);
+        snapshot.mut_metadata().set_term(term);
+
+        // Find the latest configuration state before applied index.
+        let i = match self
+            .raft_state
+            .conf_states
+            .iter()
+            .position(|cs| cs.index > applied_idx)
+        {
+            Some(i) => i,
+            None => self.raft_state.conf_states.len() - 1,
+        };
+
+        if self.raft_state.conf_states[i].in_membership_change {
+            let cs = self.raft_state.conf_states[i - 1].clone();
+            snapshot.mut_metadata().set_conf_state(cs.conf_state);
+            snapshot.mut_metadata().set_conf_state_index(cs.index);
+            let cs = self.raft_state.conf_states[i].clone();
+            snapshot.mut_metadata().set_next_conf_state(cs.conf_state);
+            snapshot.mut_metadata().set_next_conf_state_index(cs.index);
+        } else {
+            let cs = self.raft_state.conf_states[i].clone();
+            snapshot.mut_metadata().set_conf_state(cs.conf_state);
+            snapshot.mut_metadata().set_conf_state_index(cs.index);
         }
-        self.snapshot.set_data(data);
-        Ok(&self.snapshot)
+        snapshot
     }
 
     /// Discards all log entries prior to compact_index.
     /// It is the application's responsibility to not attempt to compact an index
     /// greater than RaftLog.applied.
-    pub fn compact(&mut self, compact_index: u64) -> Result<()> {
-        let offset = self.entries[0].get_index();
-        if compact_index <= offset {
-            return Err(Error::Store(StorageError::Compacted));
+    pub fn compact(&mut self, compact_index: u64) {
+        if compact_index <= self.inner_first_index() {
+            return;
```

Comment:
This used to be an error state. What changed?

Reply:
Because I don't think it's an error, and no test cases depend on the behavior. So I change it to a more simple way. As you can see all functions of `MemStorageCore` don't return `Result` now, they just exit with success or panic. I think it can save more brain when we write test cases.